ipids.go

```diff
@@ -0,0 +1,82 @@
+package zerotrace
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"sync"
+	"time"
+)
+
+var (
+	errNoMoreIds = errors.New("all IP IDs are currently in flight")
+)
+
+// ipIdPool keeps track of IP IDs that we use for traceroutes.
+type ipIdPool struct {
+	sync.Mutex // Guards ipids.
+	ipids      map[uint16]time.Time
+}
+
+func newIpIdPool() *ipIdPool {
+	return &ipIdPool{
+		ipids: make(map[uint16]time.Time),
+	}
+}
+
+// size returns the number of IP IDs that are currently in flight.
+func (s *ipIdPool) size() int {
+	s.Lock()
+	defer s.Unlock()
+
+	return len(s.ipids)
+}
+
+// borrow "borrows" an IP ID that's meant to be returned later.
+func (s *ipIdPool) borrow() (uint16, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.ipids) == math.MaxUint16 {
+		return 0, errNoMoreIds
+	}
+
+	// Start at a random index and look for available IP IDs.
+	start := uint16(rand.Intn(math.MaxUint16))
+	for id := start + 1; id != start; id++ {
```

Comment:
Does this wrap around to 0, or run off the end of the map? `rand.Intn` returns a value from the half-open interval [0, math.MaxUint16] so the maximum value possible is 65534. Adding 1 starts the search at 65535, which is the last possible key, leaving only one slot to check. The `id != start` termination check suggests it's wrapping, but wanted to confirm?

Reply:
Yes, it wraps around. Documented in 45622cf2b522c1cded57da9c88e3c1018180f909.

Reply:
Great, thanks.